ugs-core/src/com/willwinder/universalgcodesender/connection/TCPConnection.java

```diff
@@ -0,0 +1,165 @@
+/*
+	Copyright 2015-2018 Will Winder
+
+	This file is part of Universal Gcode Sender (UGS).
+
+	UGS is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	UGS is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with UGS. If not, see <http://www.gnu.org/licenses/>.
+*/
+package com.willwinder.universalgcodesender.connection;
+
+import java.io.OutputStream;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.io.BufferedReader;
+import java.net.*;
+import java.net.Socket;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * A TCP connection object implementing the connection API.
+ *
+ * @author Adam Carmicahel <carneeki@carneeki.net>
+ */
+public class TCPConnection extends AbstractConnection implements Runnable, Connection {
+
+	private String host;
+	private int port;
+
+	// General variables
+	private Socket client;
+	private BufferedReader bufIn;
+	private OutputStream bufOut;
+	private InputStreamReader inStream;
+	private Thread replyThread;
+	private ResponseMessageHandler responseMessageHandler;
+
+	@Override
+	public void setUri(String uri) {
+		try {
+			host = StringUtils.substringBetween(uri, ConnectionDriver.TCP.getProtocol(), ":");
+			port = Integer.valueOf(StringUtils.substringAfterLast(uri, ":"));
+		} catch (Exception e) {
+			throw new ConnectionException("Couldn't parse connection string " + uri, e);
+		}
+
+		if (StringUtils.isEmpty(host)) {
+			throw new ConnectionException("Empty host in connection string.");
+		}
+		if ((port < 1) || (port > 65535)) {
+			throw new ConnectionException("Please ensure port is a port number between 1 and 65535.");
+		}
+	}
+
+	@Override
+	public boolean openPort() throws Exception {
+		responseMessageHandler = new ResponseMessageHandler();
+
+		try {
+			client = new Socket(host, port);
+		} catch( BindException e) {
+			throw new ConnectionException("Could not bind a local port.");
+		} catch( NoRouteToHostException e) {
+			throw new ConnectionException("No route to host. The remote host may not be running, blocked by a firewall, or disconnected.");
+		} catch( PortUnreachableException e) {
+			throw new ConnectionException("The port is unreachable on the remote host. The server may not be running, or blocked by a firewall.");
+		} catch (SocketException e) {
+			e.printStackTrace();
```

Comment:
Is there a reason not throwing this exception further?

Reply:
IIRC SocketException is a generic of the specific types of the already listed specific exceptions. In theory, it should never be thrown, but I added it if the API introduces new instances of SocketException in the future so it would still be logged. I can remove it entirely so if it happens it just get passed up the stack like normal.